### DISCUSSION TITLE
Required fields in list are not required in graphql schema 😢

### DISCUSSION THREAD

**Discussion by tofsjonas** [keystonejs/keystone#8308] [Questions] [answered]
create a list like so:

```javascript
export const Post = list({
  access: allowAll,
  fields: {
    title: text({ validation: { isRequired: true } }),
    name: text({ validation: { isRequired: true } }),
    label: text({ validation: { isRequired: true } }),
    created_at: timestamp({ defaultValue: { kind: 'now' } }),
    updated_at: timestamp({ defaultValue: { kind: 'now' }, db: { updatedAt: true } }),
  },
})
```

If I go to <http://localhost:3000/api/graphql>, I would expect to find this in the graphql schema:

```graphql
type Post {
  id: ID!
  title: String! 
  name: String!
  label: String!
  created_at: DateTime
  updated_at: DateTime
}
input PostCreateInput {
  title: String!
  name: String!
  label: String!
  created_at: DateTime
  updated_at: DateTime
}
# etc.
```

What I am actually getting is this:

```graphql
type Post {
  id: ID!
  title: String
  name: String
  label: String
  created_at: DateTime
  updated_at: DateTime
}

input PostCreateInput {
  title: String
  name: String
  label: String
  created_at: DateTime
  updated_at: DateTime
}
# etc.
```

There are no exclamation marks 😞

**Comment by marekryb** [comment, marked as answer]
Hey there.

This is because, even though fields are required, you might restrict access to them and they will return null.
What you are looking for is [described here](https://keystonejs.com/docs/fields/text#:~:text=graphql.read.isNonNull,will%20be%20unreadable.) (note though that not all fields support this).

So you schema should look like:
```typescript
export const Post = list({
  access: allowAll,
  fields: {
    title: text({ validation: { isRequired: true }, graphql: { read: { isNonNull: true } } }),
    name: text({ validation: { isRequired: true }, graphql: { read: { isNonNull: true } } }),
    label: text({ validation: { isRequired: true }, graphql: { read: { isNonNull: true } } }),
    created_at: timestamp({ defaultValue: { kind: 'now' }, graphql: { read: { isNonNull: true } } }),
    updated_at: timestamp({ defaultValue: { kind: 'now' }, db: { updatedAt: true }, graphql: { read: { isNonNull: true } } }),
  },
})
```

EDIT: As for create input type, there is also equivalent create setting, but the output type will be something like:
```graphql
title: String! = ""
```
As far as I know there is no way to force non-null value without default value at the moment (at least for autogenerated stuff - in custom query/mutation you can do whatever you want). If required field is not provided, it will fail in runtime during validation.

**Reply by dcousens** [reply]
@marekryb please be aware this has been updated in https://github.com/keystonejs/keystone/pull/8283 (soon to be released), and now every field that has representation in the GraphQL will adhere to the `graphql.isNonNull.*` configuration.

**Comment by tofsjonas** [comment]
Thank you for your speedy reply! 🙏

I get why the ability to return null would affect the `Post` type, and

```javascript
export const Post = list({
  access: allowAll,
  fields: {
    title: text({ validation: { isRequired: true }, graphql: { create: { isNonNull: true } } }),
    name: text({ validation: { isRequired: true }, graphql: { create: { isNonNull: true } } }),
    label: text({ validation: { isRequired: true }, graphql: { create: { isNonNull: true } } }),
    created_at: timestamp({ defaultValue: { kind: 'now' } }),
    updated_at: timestamp({ defaultValue: { kind: 'now' }, db: { updatedAt: true } }),
  },
})
```

solves my immediate problem (typescript/codegen). 🥳

However, it's A LOT of code for something I think should be the default behaviour for fields with `validation: { isRequired: true }`; I will _always_ get an error message from the server if I don't post those fields.

I don't really get what is going on in #8283, will this issue be solved there?

**Comment by molomby** [comment]
Dupe of #7736?
